### PR TITLE
Account balances metrics

### PR DESCRIPTION
**Pull request by DmitriyMolch**
https://cerenetwork.atlassian.net/browse/CBI-1946

### CODE REVIEW

File: api/src/config/blockchains.js [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+const accountGroups = {
+  DDC: 'DDC',
+  BRIDGE: 'BRIDGE',
+  BRIDGE_RELAYERS: 'BRIDGE_RELAYERS',
+  DAVINCI: 'DAVINCI',
+  STATS: 'STATS'
+}
+
+const decimals = {}
+decimals[blockchainNames.CERE] = 10;
```
Comment (AndreiNavoichyk) [inline api/src/config/blockchains.js]: Decimals is per token and not per blockchain.

File: api/src/config/blockchains.js [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+
+let blockchains = (process.env.blockchains && JSON.parse(process.env.blockchains)) || [
+  {
+    "name": blockchainNames.CERE,
+    "nativeTokenSymbol": tokenSymbols.CERE,
+    "networks": [
+      {
+        "name": networkNames.DEVNET,
+        "rpcUrl": "wss://rpc.devnet.cere.network:9945",
+        "mnemonic": "",
```
Comment (AndreiNavoichyk) [inline api/src/config/blockchains.js]: Not clear what is this mnemonic for. Let's rename it to `faucetMnemonic`

File: api/src/config/blockchains.js [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+            "name": "polygon-relayer-2",
+            "minBalance": 1,
+            "group": accountGroups.BRIDGE_RELAYERS
+          },
+          {
+            "address": "0x8Ee8876d13e79b846fb6c3e5Ffe226a2e111387a",
+            "name": "polygon-erc20-handler",
+            "minBalance": 100,
+            "type": tokenTypes.ERC20,
+            "tokenSymbol": 'CN', // why it's CN and not CEERE ?
```
Comment (AndreiNavoichyk) [inline api/src/config/blockchains.js]: We will redeploy later SCs with the right token symbol.
Use here CERE so we have the right labels in the monitoring.

File: api/src/config/blockchains.js [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+      }
+    ]
+  },
+  {
+    "name": blockchainNames.POLYGON,
+    "nativeTokenSymbol": tokenSymbols.MATIC,
+    "networks": [
+      {
+        "name": networkNames.TESTNET,
+        "rpcUrl": "https://polygon-mumbai.infura.io/v3/ef1894c45068455680d1b08a35a6e83b",
```
Comment (AndreiNavoichyk) [inline api/src/config/blockchains.js]: Infura urls should be removed

File: api/src/config/blockchains.js [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+            "type": tokenTypes.ERC20,
+            "tokenSymbol": 'CN', // why it's CN and not CEERE ?
+            "erc20TokenAddress": "0xd111d479e23A8342A81ad595Ea1CAF229B3528c3",
+            "group": accountGroups.BRIDGE
+          }
+        ]
+      },
+      {
+        "name": networkNames.MAINNET,
+        "rpcUrl": "https://polygon-mainnet.infura.io/v3/b02e9d9902654c29b15508f943aab443",
```
Comment (AndreiNavoichyk) [inline api/src/config/blockchains.js]: Infura urls should be removed

File: api/src/lib/utils/index.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
+const { BigNumber } = require('ethers');
```
Comment (AndreiNavoichyk) [inline api/src/lib/utils/index.js]: We don't need ethers only for `BigNumber`. BN.js uses by web3.

File: api/src/lib/utils/index.js [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+function convertSmallToBigCoins (smallCoins, decimals) {
```
Comment (AndreiNavoichyk) [inline api/src/lib/utils/index.js]: 1. Check web3 for this method
2. If no, rename to `toDecimal`.

File: api/src/services/cacheService.js [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+                case tokenTypes.ERC20:
+                  balance = await ethNetworkService.getErc20Balance({
+                    blockchain: blockchain.name,
+                    network: network.name,
+                    erc20TokenAddress: account.erc20TokenAddress,
+                    address: account.address
+                  });
+                  break;
+              }
+              break;
```
Comment (AndreiNavoichyk) [inline api/src/services/cacheService.js]: Add log.warn in `default:` section when blockchain is unknown.

File: api/src/services/cacheService.js [hunk 9]
```diff
@@ -0,0 +1,71 @@
+const { blockchainNames, tokenTypes } = require("../config/blockchains");
```
Comment (AndreiNavoichyk) [inline api/src/services/cacheService.js]: This service has too specific logic to be named `cacheService`. Let's rename it to `accountsBalancesService`.

File: api/src/services/cacheService.js [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+const { blockchains, cacheIntervalMs } = require("../config");
+const cereNetworkService = require('./cereNetworkService');
+const ethNetworkService = require('./ethNetworkService');
+
+let cache = {
+  accounts: []
+};
+
+async function  requestAccounts() {
```
Comment (AndreiNavoichyk) [inline api/src/services/cacheService.js]: Be more specific. Rename to `requestAccountsBalances`

File: api/src/services/cereNetworkService.js [hunk 11]
```diff
@@ -1,15 +1,26 @@
...
+    });
+    promises.push(promise());
+  });
+  const res = await Promise.all(promises);
+  res.forEach(network => {
+    networkParams.set(network.name, network.rpc);
+  });
 
+  // Deprecated: init networks from NETWORKS config
+  // ToDo: remove it in the future from code and devops configs
```
Comment (AndreiNavoichyk) [inline api/src/services/cereNetworkService.js]: Let's rename `devops configs` to `deployment configs`

File: api/src/services/cereNetworkService.js [hunk 12]
```diff
@@ -1,15 +1,26 @@
...
-} = process.env;
+const { cereTypes, blockchains } = require("../config");
+const { NETWORKS } = process.env;
 
 const networkParams = new Map();
 
 async function init() {
-  networkParams.set(CERE_MAINNET, await initNetwork(CERE_WS_PROVIDER_URL || CERE_MAINNET_WS_PROVIDER_URL_DEFAULT, MNEMONIC_EMPTY));
+  const promises = [];
+  const cere = blockchains[0];
```
Comment (AndreiNavoichyk) [inline api/src/services/cereNetworkService.js]: We can not be sure. Please, update the logic.

File: api/src/services/cereNetworkService.js [hunk 13]
```diff
@@ -1,15 +1,26 @@
...
+  cere.networks.forEach(network => {
+    const promise = async() => ({
+      name: network.name,
+      rpc: await initNetwork(network.rpcUrl, network.mnemonic)
```
Comment (AndreiNavoichyk) [inline api/src/services/cereNetworkService.js]: `mnemonic` is in NETWORKS that is taken from secrets. You should use it.

File: api/src/services/cereNetworkService.js [hunk 14]
```diff
@@ -50,16 +61,15 @@ function initFaucet(mnemonic) {
...
 module.exports = {
+  init,
   supportsNetwork: (network) => {
     return networkParams.has(network.toUpperCase());
   },
   getBalance: async (network, address) => {
     const { api, _ } = networkParams.get(network.toUpperCase());
     const { nonce, data: balance } = await api.query.system.account(address);
-    return balance.free;
+    return balance.free.toString();
```
Comment (AndreiNavoichyk) [inline api/src/services/cereNetworkService.js]: Let's return BigNumber with a value converted to decimal.

File: api/src/services/cereTokensService.js [hunk 15]
```diff
@@ -1,24 +1,24 @@
-const { decimal, CERE_MAINNET } = require("../constants/config");
+const { blockchainNames, networkNames, decimals } = require("../config/blockchains");
 const cereNetworkService = require('./cereNetworkService');
 const ethNetworkService = require('./ethNetworkService');
 const { ETHEREUM_CERE_LOCKED_ADDRESSES } = process.env;
 
 async function getTotalSupplyInternal(network) {
   const totalSupply = await cereNetworkService.getTotalSupply(network);
-  return +totalSupply / 10 ** decimal;
+  return +totalSupply / 10 ** decimals[blockchainNames.CERE];
```
Comment (AndreiNavoichyk) [inline api/src/services/cereTokensService.js]: We should use BigNumber here as well.

File: api/src/services/ethNetworkService.js [hunk 16]
```diff
@@ -1,16 +1,71 @@
...
+}
+
+function getProviderId(blockchain, network) {
+  return  `${blockchain}_${network}`;
+}
 
 module.exports = {
-  getCEREBalanceOf: async (address) => {
+  init,
+  getCEREBalanceOf: async (blockchain, network, address) => {
```
Comment (AndreiNavoichyk) [inline api/src/services/ethNetworkService.js]: Let's get rid of this method and use `getErc20Balance` instead.

File: api/src/services/metricsService.js [hunk 17]
```diff
@@ -0,0 +1,24 @@
+const prom = require('prom-client');
+const { decimals } = require('../config/blockchains');
+const cacheService = require('./cacheService');
+const { convertSmallToBigCoins } = require('../lib/utils')
+
+const accountBalancesMetric = new prom.Gauge({
```
Comment (AndreiNavoichyk) [inline api/src/services/metricsService.js]: rename to `account**s**BalancesMetric`

File: api/src/services/metricsService.js [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+  name: 'accounts_balances',
+  help: 'internal accounts balances amount',
```
Comment (AndreiNavoichyk) [inline api/src/services/metricsService.js]: Let's remove `internal`. We are monitoring internal and external.

File: api/src/services/cacheService.js (hunk 9 above)
Comment (DmitriyMolch) [reply api/src/services/cacheService.js]: The idea was to create a common cache service. In the future, I planned to extend it for all caching purposes and not only for accounts balances. By the way, we cache all accounts and not just their balances.

File: api/src/services/cacheService.js (hunk 10 above)
Comment (DmitriyMolch) [reply api/src/services/cacheService.js]: We request here not `balances` but the whole `account`.

File: api/src/services/cereNetworkService.js (hunk 14 above)
Comment (DmitriyMolch) [reply api/src/services/cereNetworkService.js]: I suggest using everywhere strings as an interface. `BigNumber` is a wrap required for mathematical operations and string is a more universal format that can be used to create `BigNumber`, send it to the client, and store it in the database.

File: api/src/services/cereTokensService.js (hunk 15 above)
Comment (DmitriyMolch) [reply api/src/services/cereTokensService.js]: I suggest using here `BigNumber` just for mathematical operations and return a string in the end.

File: api/src/services/cereNetworkService.js (hunk 13 above)
Comment (DmitriyMolch) [reply api/src/services/cereNetworkService.js]: Ok, I will create a  possibility to configure the mnemonic from the config file via env variables..

File: api/src/services/cacheService.js (hunk 9 above)
Comment (AndreiNavoichyk) [reply api/src/services/cacheService.js]: Discussed to split into 2 services to respect S from SOLID.

File: api/src/services/cacheService.js (hunk 10 above)
Comment (AndreiNavoichyk) [reply api/src/services/cacheService.js]: Will be renamed after splitting.

File: api/src/services/cereNetworkService.js (hunk 14 above)
Comment (AndreiNavoichyk) [reply api/src/services/cereNetworkService.js]: Discussed. Let's use BN.js

File: api/src/services/cereNetworkService.js (hunk 13 above)
Comment (AndreiNavoichyk) [reply api/src/services/cereNetworkService.js]: Discussed. Mnemonic will be taken from NETOWRKS and task for migration created.

File: api/src/config/blockchains.js (hunk 4 above)
Comment (DmitriyMolch) [reply api/src/config/blockchains.js]: Why it's bad that somebody will start to use URLs from development environments? Removing them from config requires each time to search them somewhere.

File: api/src/config/blockchains.js (hunk 1 above)
Comment (DmitriyMolch) [reply api/src/config/blockchains.js]: resolved

File: api/src/config/blockchains.js (hunk 2 above)
Comment (DmitriyMolch) [reply api/src/config/blockchains.js]: resolved

File: api/src/config/blockchains.js (hunk 3 above)
Comment (DmitriyMolch) [reply api/src/config/blockchains.js]: resolved

File: api/src/config/blockchains.js (hunk 4 above)
Comment (DmitriyMolch) [reply api/src/config/blockchains.js]: resolved

File: api/src/config/blockchains.js (hunk 5 above)
Comment (DmitriyMolch) [reply api/src/config/blockchains.js]: resolved

File: api/src/lib/utils/index.js (hunk 6 above)
Comment (DmitriyMolch) [reply api/src/lib/utils/index.js]: resolved

File: api/src/lib/utils/index.js (hunk 7 above)
Comment (DmitriyMolch) [reply api/src/lib/utils/index.js]: resolved

File: api/src/services/cacheService.js (hunk 8 above)
Comment (DmitriyMolch) [reply api/src/services/cacheService.js]: resolved

File: api/src/services/cacheService.js (hunk 9 above)
Comment (DmitriyMolch) [reply api/src/services/cacheService.js]: resolved

File: api/src/services/cacheService.js (hunk 10 above)
Comment (DmitriyMolch) [reply api/src/services/cacheService.js]: resolved

File: api/src/services/cereNetworkService.js (hunk 11 above)
Comment (DmitriyMolch) [reply api/src/services/cereNetworkService.js]: resolved

File: api/src/services/cereNetworkService.js (hunk 12 above)
Comment (DmitriyMolch) [reply api/src/services/cereNetworkService.js]: resolved

File: api/src/services/cereNetworkService.js (hunk 13 above)
Comment (DmitriyMolch) [reply api/src/services/cereNetworkService.js]: resolved

File: api/src/services/cereNetworkService.js (hunk 14 above)
Comment (DmitriyMolch) [reply api/src/services/cereNetworkService.js]: resolved

File: api/src/services/cereTokensService.js (hunk 15 above)
Comment (DmitriyMolch) [reply api/src/services/cereTokensService.js]: resolved

File: api/src/services/ethNetworkService.js (hunk 16 above)
Comment (DmitriyMolch) [reply api/src/services/ethNetworkService.js]: resolved

File: api/src/services/metricsService.js (hunk 17 above)
Comment (DmitriyMolch) [reply api/src/services/metricsService.js]: resolved

File: api/src/services/metricsService.js (hunk 18 above)
Comment (DmitriyMolch) [reply api/src/services/metricsService.js]: resolved

File: api/package.json [hunk 19]
```diff
@@ -16,6 +16,7 @@
     "body-parser": "^1.19.0",
     "cors": "^2.8.5",
     "dotenv": "^10.0.0",
+    "ethers": "^5.6.4",
```
Comment (AndreiNavoichyk) [inline api/package.json]: Remove it from the dependencies

File: api/src/config/blockchains.js [hunk 20]
```diff
@@ -0,0 +1,309 @@
...
+  BRIDGE_RELAYERS: 'BRIDGE_RELAYERS',
+  DAVINCI: 'DAVINCI',
+  STATS: 'STATS'
+}
+
+const decimals = {}
+decimals[tokenSymbols.CERE] = new BN(10);
+decimals[tokenSymbols.MATIC] = new BN(18);
+decimals[tokenSymbols.ETH] = new BN(18);
+decimals[tokenSymbols.USDC] = new BN(18);
```
Comment (AndreiNavoichyk) [inline api/src/config/blockchains.js]: Where did you take 18? USDC's decimals are 6.

Comment (AndreiNavoichyk) [inline api/src/config/blockchains.js]: Please, create a task to get decimals from code and add its ID here with ToDo comment

File: api/src/services/accountsService.js [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+        const promise = async () => {
+          let balance;
+          switch (blockchain.name) {
+            case blockchainNames.CERE:
+              switch(account.type) {
+                case tokenTypes.NATIVE:
+                  balance = await cereNetworkService.getBalance(network.name, account.address);
+                  break;
+              }
+              break;
```
Comment (AndreiNavoichyk) [inline api/src/services/accountsService.js:22]: Let's also add warn logs when the tokenType is unknown to all blockchains

File: api/src/services/cereNetworkService.js (hunk 12 above)
Comment (AndreiNavoichyk) [reply api/src/services/cereNetworkService.js]: How? I still see that you suppose that cere blockchain has index `0`

File: api/src/config/blockchains.js (hunk 20 above)
Comment (DmitriyMolch) [reply api/src/config/blockchains.js]: https://cerenetwork.atlassian.net/browse/CBI-1971

File: api/src/services/cereNetworkService.js (hunk 12 above)
Comment (DmitriyMolch) [reply api/src/services/cereNetworkService.js]: resolved

File: api/package.json (hunk 19 above)
Comment (DmitriyMolch) [reply api/package.json]: resolved

File: api/src/config/blockchains.js (hunk 20 above)
Comment (DmitriyMolch) [reply api/src/config/blockchains.js]: resolved

Comment (DmitriyMolch) [reply api/src/config/blockchains.js]: resolved

File: api/src/services/accountsService.js (hunk 21 above)
Comment (DmitriyMolch) [reply api/src/services/accountsService.js:22]: resolved

**Review by AndreiNavoichyk — APPROVED** [review]

**Pull request merged**